llmware/parsers.py

```diff
@@ -3066,6 +3066,151 @@ def basic_library_duplicate_check(self, fn):
+
+        #   if found in mapping dict, then will over-write
+        reserved_keys = ["text", "doc_ID", "block_ID"]
+
+        rejected_rows = []
+
+        if not mapping_dict:
+            raise LLMWareException(message="Parsing of a configured CSV file requires a mapping dictionary so that "
+                                           "the table attributes can be properly mapped.")
+
+        if not cols:
+            raise LLMWareException(message="Parsing of a configured CSV file requires a defined column structure and "
+                                           "a specified number of columns to ensure accurate mapping.")
+
+        # file type
+        ft = fn.split(".")[-1].lower()
+        if ft == "tsv":
+            delimiter="\t"
+
+        # will iterate through csv file
+        input_file = os.path.join(fp, fn)
+
+        f = open(input_file, 'r', encoding='utf-8-sig',errors='ignore')
+        filedata = f.readlines()
+
+
+        output = []
+        for line in filedata:
+            output.append(line.split(delimiter))
+        f.close()
+        del filedata
+
+        added_row_count = 0
+        total_row_count = 0
+        added_doc_count = 0
+
+        for i, rows in enumerate(output):
+
+            text = ""
+            doc_id = None
+            block_id = None
+            metadata = {}
+
+            if len(rows) != cols:
+                bad_entry = {"index": i, "row": rows}
+                rejected_rows.append(bad_entry)
+
+            else:
+                # confirmed that row has the correct number of entries
+
+                for keys, values in mapping_dict.items():
+
+                    if keys == "text":
+                        if mapping_dict["text"] < len(rows):
+                            text = rows[mapping_dict["text"]]
+
+                    if keys == "doc_ID":
+                        if mapping_dict["doc_ID"] < len(rows):
+                            doc_id = rows[mapping_dict["doc_ID"]]
+
+                    if keys == "block_ID":
+                        if mapping_dict["block_ID"] < len(rows):
+                            block_id = rows[mapping_dict["block_ID"]]
+
+                    if keys not in reserved_keys:
+                        if values < len(rows):
+                            metadata.update({keys:rows[values]})
+
+            if text.strip():
+
+                meta = {"author": "", "modified_date": "", "created_date": "", "creator_tool": ""}
+                coords_dict = {"coords_x": 0, "coords_y": 0, "coords_cx": 0, "coords_cy": 0}
+
+                # note: if using SQL-based DB, then will save the metadata as a text string
+                if LLMWareConfig().get_config("collection_db") in ["sqlite", "postgres"]:
+                    metadata = str(metadata)
+
+                new_row_entry = ("text", "custom_csv", (1, 0), total_row_count, "", "", fn,
+                                 text, text, "", "", text, text, "", text, "", "", metadata, "", "")
```

Comment:
This here is a problem, b/c it is hard to understand what each column is for. While this i not peculiar to this PR, we should think about adding an ``Enum`` for this.